src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,381 @@
+{
+	init();
+	k=k_;
+	set_distance(d);
+	use_kmeanspp=use_kmpp;
+}
+
+CKMeansBase::CKMeansBase(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i)
+: CDistanceMachine()
+{
+	init();
+	k = k_i;
+	set_distance(d_i);
+	set_initial_centers(centers_i);
+}
+
+CKMeansBase::~CKMeansBase()
+{
+}
+
+void CKMeansBase::set_initial_centers(SGMatrix<float64_t> centers)
+{
+ 	CDenseFeatures<float64_t>* lhs=((CDenseFeatures<float64_t>*) distance->get_lhs());
+	dimensions=lhs->get_num_features();
+	REQUIRE(centers.num_cols == k,
+			"Expected %d initial cluster centers, got %d", k, centers.num_cols);
+	REQUIRE(centers.num_rows == dimensions,
+			"Expected %d dimensionional cluster centers, got %d", dimensions, centers.num_rows);
+	mus_initial = centers;
+	SG_UNREF(lhs);
+}
+
+void CKMeansBase::set_random_centers()
+{
+	mus.zero();
+	CDenseFeatures<float64_t>* lhs=
+		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+	int32_t lhs_size=lhs->get_num_vectors();
+	
+	SGVector<int32_t> temp=SGVector<int32_t>(lhs_size);
+	SGVector<int32_t>::range_fill_vector(temp, lhs_size, 0);
+	CMath::permute(temp);
+
+	for (int32_t i=0; i<k; i++)
+	{
+		const int32_t cluster_center_i=temp[i];
+		SGVector<float64_t> vec=lhs->get_feature_vector(cluster_center_i);
+
+		for (int32_t j=0; j<dimensions; j++)
+			mus(j,i)=vec[j];
+
+		lhs->free_feature_vector(vec, cluster_center_i);
+	}
+
+	SG_UNREF(lhs);
+
+}
+
+void CKMeansBase::compute_cluster_variances()
+{
+	/* compute the ,,variances'' of the clusters */
+	for (int32_t i=0; i<k; i++)
+	{
+		float64_t rmin1=0;
+		float64_t rmin2=0;
+
+		bool first_round=true;
+
+		for (int32_t j=0; j<k; j++)
+		{
+			if (j!=i)
+			{
+				int32_t l;
+				float64_t dist = 0;
+
+				for (l=0; l<dimensions; l++)
+				{
+					dist+=CMath::sq(
+							mus.matrix[i*dimensions+l]
+									-mus.matrix[j*dimensions+l]);
+				}
+
+				if (first_round)
+				{
+					rmin1=dist;
+					rmin2=dist;
+					first_round=false;
+				}
+				else
+				{
+					if ((dist<rmin2) && (dist>=rmin1))
+						rmin2=dist;
+
+					if (dist<rmin1)
+					{
+						rmin2=rmin1;
+						rmin1=dist;
+					}
+				}
+			}
+		}
+
+		R.vector[i]=(0.7*CMath::sqrt(rmin1)+0.3*CMath::sqrt(rmin2));
+	}
+}
+
+void CKMeansBase::initialize_training(CFeatures* data)
+{
+	REQUIRE(distance, "Distance is not provided")
+	REQUIRE(distance->get_feature_type()==F_DREAL, "Distance's features type (%d) should be of type REAL (%d)")
+	
+	if (data)
+		distance->init(data, data);
+
+	CDenseFeatures<float64_t>* lhs=
+		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+
+	REQUIRE(lhs, "Lhs features of distance not provided");
+	int32_t lhs_size=lhs->get_num_vectors();
+	dimensions=lhs->get_num_features();
+	const int32_t centers_size=dimensions*k;
+
+	REQUIRE(lhs_size>0, "Lhs features should not be empty");
+	REQUIRE(dimensions>0, "Lhs features should have more than zero dimensions");
+
+	/* if kmeans++ to be used */
+	if (use_kmeanspp)
+	{
+#ifdef HAVE_LINALG_LIB
```

Comment:
what happens if we dont have linalg lib?
i mean in that case we should at least warn the user that kmeans++ wont be used or something...